### ISSUE TITLE
[Flake] Crash on unit test with Windows+CMake build

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#3933]
Looks like a memory management error, but could be linking against the wrong CRT:

https://source.cloud.google.com/results/invocations/7d573da1-56d4-451f-b42a-73a7241eba8f/targets

```
29/32 Test #29: internal_background_threads_impl_test ...***Exception: SegFault  0.29 sec
[==========] Running 3 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from CustomerSuppliedBackgroundThreads
[ RUN      ] CustomerSuppliedBackgroundThreads.LifecycleNoShutdown
[       OK ] CustomerSuppliedBackgroundThreads.LifecycleNoShutdown (35 ms)
[ RUN      ] CustomerSuppliedBackgroundThreads.SharesCompletionQueue
[       OK ] CustomerSuppliedBackgroundThreads.SharesCompletionQueue (15 ms)
[----------] 2 tests from CustomerSuppliedBackgroundThreads (50 ms total)

[----------] 1 test from AutomaticallyCreatedBackgroundThreads
[ RUN      ] AutomaticallyCreatedBackgroundThreads.IsActive
[       OK ] AutomaticallyCreatedBackgroundThreads.IsActive (4 ms)
[----------] 1 test from AutomaticallyCreatedBackgroundThreads (4 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 2 test suites ran. (56 ms total)
[  PASSED  ] 3 tests.
minkernel\crts\ucrt\src\appcrt\heap\debug_heap.cpp(904) : Assertion failed: _CrtIsValidHeapPointer(block)
```

**Comment by coryan** [comment]
I was able to capture a stack trace, no text, just the visual version:

![#3933-stack-trace](https://user-images.githubusercontent.com/6241635/81085048-acad6380-8ec4-11ea-96de-dea0fd46ea3d.png)


I doubt this is our code. It seems the problem is during the gRPC shutdown, which reminds me of #2249 and #3273. The code in gRPC doing the shutdown seems like a double delete waiting to happen:


![#3933-grpc-crash](https://user-images.githubusercontent.com/6241635/81085849-c00cfe80-8ec5-11ea-84ab-2c45a686a56b.png)

**Comment by coryan** [comment]
I just noticed that the main thread is shutting down the winapi when this happens. That cannot be good:

![#3933-main-stack-trace](https://user-images.githubusercontent.com/6241635/81113514-5b19ce80-8eee-11ea-9417-42e9ee9fb85b.png)

**Comment by devjgm** [comment]
As you point out, it may not be our code. But crashes during shutdown make me think of globals that aren't trivially destructible. Not sure if any of those are lurking around here.

**Comment by coryan** [comment]
I think I have a workaround: compile the code in release mode.  I was just able to run the tests 2,000 times, where normally it crashes before completing 100 runs.

And I think I know *why* this workaround works (at least for some of the problems):

The STL for MSVC+Debug is very different from the Release version, it includes code like this:

https://github.com/microsoft/STL/blob/e4bc00e70cbb539f90b803a64a31f0259e21f28e/stl/inc/xmemory#L1206

which acquires a lock:

https://github.com/microsoft/STL/blob/e4bc00e70cbb539f90b803a64a31f0259e21f28e/stl/src/xlock.cpp#L70

but that is a global lock:

https://github.com/microsoft/STL/blob/e4bc00e70cbb539f90b803a64a31f0259e21f28e/stl/src/xlock.cpp#L19

which may have been shutdown because all this code is running after `main()` exits.

In short: using the STL after `main()` with MSVC+Debug may be UB, so it is a big no-no.

**Comment by coryan** [comment]
I think this has been reported to gRPC as  grpc/grpc#22479

**Comment by coryan** [comment]
I am marking this as blocked on an external issue, will disable the Debug build for MSVC, and move on.  If it (or something very similar) does not repeat for 90 days we will close it as a Flake that no longer happens.

**Comment by coryan** [comment]
The quickstart builds also suffer from this problem:

https://source.cloud.google.com/results/invocations/82f385f7-c2fd-4e72-ae14-e26e52f19c4b

**Comment by coryan** [comment]
https://source.cloud.google.com/results/invocations/0e2058c6-640b-4201-b63c-631aaa38279f/details

**Comment by coryan** [comment]
This is not on Windows, but similar, a crash after `main()` exits with a gRPC programs:

https://source.cloud.google.com/results/invocations/187291a7-c02e-405c-b613-88d7f05d97d9/details

**Comment by coryan** [comment]
A potential (but really ugly) workaround:

https://stackoverflow.com/questions/60849381/grpc-access-violation-on-client

**Comment by coryan** [comment]
https://source.cloud.google.com/results/invocations/93f031eb-3537-4c59-a031-3bd53336e403/targets

**Comment by coryan** [comment]
https://source.cloud.google.com/results/invocations/d1a97494-5239-46ee-95fd-7816519d0af4

**Comment by devjgm** [comment]
Another one https://g3c.corp.google.com/results/invocations/36942350-0b32-4e45-848a-b0ac220cd8ba/targets/cloud-devrel%2Fclient-libraries%2Fcpp%2Fgoogle-cloud-cpp%2Fwindows%2Fquickstart-cmake-static/log

**Comment by coryan** [comment]
They still happen, I sent a PR to gRPC that should fix these:

https://source.cloud.google.com/results/invocations/32cfa3e7-df6e-4342-b5cd-023c37674fe5

**Comment by coryan** [comment]
Last recorded repeat was more than 90d ago, and there was a fix introduced since then. Closing.
